Content.Shared/Clothing/EntitySystems/SharedChameleonClothingSystem.cs

```diff
@@ -0,0 +1,67 @@
+﻿using Content.Shared.Clothing.Components;
+using Content.Shared.Inventory;
+using Content.Shared.Item;
+using Content.Shared.Tag;
+using Robust.Shared.Prototypes;
+
+namespace Content.Shared.Clothing.EntitySystems;
+
+public abstract class SharedChameleonClothingSystem : EntitySystem
+{
+    [Dependency] private readonly IComponentFactory _factory = default!;
+    [Dependency] private readonly IPrototypeManager _proto = default!;
+    [Dependency] private readonly SharedItemSystem _itemSystem = default!;
+    [Dependency] private readonly ClothingSystem _clothingSystem = default!;
+
+    // Updates chameleon visuals and meta information.
+    // This function is called on a server after user selected new outfit.
+    // And after that on a client after state was updated.
+    // This 100% makes sure that server and client have exactly same data.
+    protected virtual void UpdateVisuals(EntityUid uid, ChameleonClothingComponent component)
+    {
+        if (string.IsNullOrEmpty(component.SelectedId) ||
+            !_proto.TryIndex(component.SelectedId, out EntityPrototype? proto))
+            return;
+
+        // copy name and description
+        var meta = MetaData(uid);
+        meta.EntityName = proto.Name;
+        meta.EntityDescription = proto.Description;
+
+        // item sprite logic
+        if (TryComp(uid, out ItemComponent? item) &&
+            proto.TryGetComponent(out ItemComponent? otherItem, _factory))
+        {
+            _itemSystem.CopyVisuals(uid, otherItem, item);
+        }
+
+        // clothing sprite logic
+        if (TryComp(uid, out SharedClothingComponent? clothing) &&
+            proto.TryGetComponent("Clothing", out SharedClothingComponent? otherClothing))
+        {
+            _clothingSystem.CopyVisuals(uid, otherClothing, clothing);
+        }
+    }
+
+    /// <summary>
+    ///     Check if this entity prototype is valid target for chameleon item.
+    /// </summary>
+    public bool IsValidTarget(EntityPrototype proto, SlotFlags chameleonSlot = SlotFlags.NONE)
+    {
+        // check if entity is valid
+        if (proto.Abstract || proto.NoSpawn)
+            return false;
+
+        // check if it isn't marked as invalid chameleon target
+        if (proto.TryGetComponent(out TagComponent? tags, _factory) && tags.Tags.Contains("IgnoreChameleon"))
+            return false;
```

Comment:
Tag needs to be whitelist instead of blacklist.

Reply:
You sure? I think blacklist is really fine here as 99% of clothing should be valid, that's the point of the suit.

Reply:
Yeah, whitelist will be real pain, because there are hundreds of potential targets that chameleon can mimic.

Reply:
> You sure? I think blacklist is really fine here as 99% of clothing should be valid, that's the point of the suit.

I checked with Remie at the time and we both agreed on a whitelist.

The pain is every contributor is going to have issues if they add something like say, clothingcomponent to a mob and now they can show that mob in the slot, because there's implicit behaviour now being done on another system you may not be aware of.

Reply:
> I doubt you need more than like... 20 abstract prototypes tagged?

Tags doesn't inherit well. So if any clothing has some additional tag, you need to go explicitly to their child and set `ChameleonWhitelist` tag. Or it will be overwritten by newly added tags.

> add something like say, clothingcomponent to a mob

Well, the actual problem is that right now codebase doesn't distinguish entities that are "fitting into inventory slot" and entities that are actual clothing. For example fireaxe or guitar are also consider to be clothing when they are clearly not.

I can suggest other solution. I can set whitelist of prototype ids explicitly in some `ChameleonWhitelistPrototype` and save id of this prototype in component. This will give explicit control over what chameleon clothing can mimic, but also add a lot of boilerplate.

Reply:
I think just putting the whitelist tag on all of the base clothing prototypes OR the `Clothing` prototype should probably work as stuff that shouldn't show up since rats don't inherit from it anyway?

I had a look and there's only 7 prototypes in the clothing folder with the Tag component on them.

Reply:
I added `WhitelistChameleon` to `Clothing` prototype. Which means that almost every clothing is can be target for chameleon again. If `Clothing` child don't want to be chameleon target, usually it should something like this.

```
    - type: Tag
      tags: [] # ignore "WhitelistChameleon" tag
```

Reply:
if we go ahead with not using tagcomp for most stuff then this is just going to become even more hell. as it stands right now this basically -is- a blacklist because stuff needs to specify the empty tag anyway and it automatically inherits the whitelist. theres literally no point to making it a whitelist @metalgearsloth

Reply:
if you want to avoid the 'clothing on mob' thing and make it not CBT with inheritance you could just make this a field on clothingcomponent i guess?

Reply:
> not to mention that now clothing which specifies other tags will be unavailable for chameleon unless people remember to put it there

Blacklist has meta issues:
- forbidden items get included (e.g. cat ears)
- Unremoveable items like hardsuit helmets get included

Maybe it should just clone like the agent card because I foresee github issues being made and it's gonna get the same meta abuse that centcomm uniforms did on uplink before we removed them (which are also included here).

Also unrelated but the UI doesn't align with the actual sprite you select.
![image](https://user-images.githubusercontent.com/31366439/187674800-1e18efbf-8163-4404-9e68-52f2b07aee51.png)

Plus the mask being pulled up / down state isn't synced.

Reply:
> if you want to avoid the 'clothing on mob' thing and make it not CBT with inheritance you could just make this a field on clothingcomponent

I don't think that it should be done this way. Some forks may not use chameleon clothing. So it will be useless field that breaks muh SOLID

I agree that current solution with tag overriding is total hell. Of course I can go and mark with tag every single clothing prototype child directly (not their parents), but this is even worse. Because if parent want to add some tag, you need to go for each child proto and add it too with all previously inherited tags manually.

Reply:
> forbidden items get included (e.g. cat ears)

Intended, unless it marked as blacklist for chameleon directly

> Unremoveable items like hardsuit helmets get included

Well, right now they are marked as no spawn, so they don't appear in search. But anyway they doesn't have unremovable behaviour, chameleon just mimic their appearance. Can't see why would it be an issue.

> Maybe it should just clone like the agent card because I foresee github issues being made and it's gonna get the same meta abuse that centcomm uniforms did on uplink before we removed them (which are also included here).

Before identity system and crew manifest that would be a problem. Right now not.

I can add blacklist for chameleon clothing as separate prototype and don't mess with proto inheritance or tag at all. It will be just a list of all blacklisted or whitelisted prototypes.

Reply:
Any ideas where should I move this whitelist issue? Current options are:
- Return it to blacklist tag
- Use current implementation with whitelist tag
- Move whitelist to separate prototype and set all clothing entities manually